### ISSUE TITLE
How to: Timeline ranged bar charts

### ISSUE THREAD

**Issue by leohickstein-ab** [recharts/recharts#4038]
## Do you want to request a feature or report a bug?
Feature

## What is the current behavior?
Recharts doesn't currently support timelime/ranged bar charts

## What is the expected behavior?
I want to build a date ranged bar chart similar to the example below (although data and look and feel will be different) but it doesn't seem Recharts has any examples in the documentation and it's being really hard to find examples online as well.

![nCjQc](https://github.com/recharts/recharts/assets/87674812/86ae80c5-3025-4601-a8fc-9a616bf0c70f)

The bars need to be stacked in the horizontal, and the Y Axis will be month-based.

Current code attempt: [https://codesandbox.io/p/sandbox/condescending-cookies-qt59r6](https://codesandbox.io/p/sandbox/condescending-cookies-qt59r6)

I think I almost got it working, except that the bars should be draw as a range and be in sync with the months from Y Axis.

## Which versions of Recharts, and which browser / OS are affected by this issue? Did this work in previous versions of Recharts?
Latest

**Comment by nikolasrieble** [comment]
Great, this could be a wonderful addition to storybook examples. ❤️

**Comment by AnujSharma141** [comment]
hey.. isn't this related to [https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/story/examples-barchart--stacked-with-error-bar](url) ... we have horizontal stacked bars with errorbars there.. we can achieve this by modifying data and removing the errorbar simply 
![image](https://github.com/recharts/recharts/assets/16597292/76bf6a86-f48e-4d1c-934b-96b1d9898dfd)

**Comment by leohickstein-ab** [comment]
@ckifer, that's correct. For a simple stacked bar chart, there is an example on the Recharts website [link](https://recharts.org/en-US/examples/StackedBarChart).

In my case, the requirements are a bit different as each bar has a starting and an ending date. These bars are drawn according to the Y-axis, which is typically based on time (hour/day/month/year), similar to the example in my image (which I got from Apex Charts).

**Comment by leohickstein-ab** [comment]
Has anyone done this before and would not mind sharing an example here? I would greatly appreciate it.

**Comment by ckifer** [comment]
I don't think it is currently possible for the way that you would want unless @nikolasrieble has any ideas. 

Stacked chart wouldn't align the data to a certain datapoint, only extend the current datapoint by a certain amount. 

With a regular bar you'd end up with groups/series. 

There may be some hack to do it but I'm not sure

**Comment by nikolasrieble** [comment]
It is certainly possible, and I know @kinmat has implemented something very similar, but I can not guarantee any capacity being invested here. 

If I remember correctly the way this was implemented was with custom bar shapes, which internally used rectangles that are stacked.

**Comment by ALEEF02** [comment]
+1 Bump!

**Comment by CollinsSpencer** [comment]
+1 from me too.

I don't think it's possible to get this working 100% today (with latest @2.12.7)

I made a sandbox with as close to a working timeline chart as I could create. [Sandbox](https://codesandbox.io/p/sandbox/recharts-timeline-hv662l?file=%2Fsrc%2FTimelineGraphFlatData.tsx%3A101%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clx2r715m00h8356ia1cah07v%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clx2r715m00h5356imfib0ulh%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clx2r715m00h7356i4v2vg23w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clx2r715m00h6356i0vi2dsma%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B55.51970703722449%252C44.48029296277551%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clx2r715m00h5356imfib0ulh%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx238ddj0001356izbfw8awv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx2941ya0002356i4w7c8npk%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fnode_modules%252F.store%252Frecharts%25402.12.7%252Fnode_modules%252Frecharts%252Ftypes%252Fcartesian%252FBar.d.ts%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx4r50v60002356i6x2upqsh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A101%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A101%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FTimelineGraphFlatData.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clx2r715m00h5356imfib0ulh%2522%252C%2522activeTabId%2522%253A%2522clx4r50v60002356i6x2upqsh%2522%257D%252C%2522clx2r715m00h6356i0vi2dsma%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx238ddk0004356i31lcskmf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clx2r715m00h6356i0vi2dsma%2522%252C%2522activeTabId%2522%253A%2522clx238ddk0004356i31lcskmf%2522%257D%252C%2522clx2r715m00h7356i4v2vg23w%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clx2r715m00h7356i4v2vg23w%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

![image](https://github.com/recharts/recharts/assets/16708427/58a1bd5e-0f18-4803-a280-a841a1be8477)

While it looks correct, the `Tooltip` doesn't display the correct data, making it non-functional. Also, it required a hacky/duplicate `YAxis`, a custom `Cell`s for coloring, and a custom `Legend` `payload`.

Here are some of my observations when trying to get it working:

- It would be nice to get this working with a single `YAxis`.
   - `allowDuplicatedCategory={false}` only displays the first data point in each category. I had to use a hidden, numeric `YAxis`. With multiple bars on the same y coordinate, this breaks the `Tooltip`. 
- Single vs multiple `Bar`s?
   - Using a single `Bar` means that there is only a single color by default and a single entry in the `Legend`. However, it displays the bars correctly.
   - Using multiple `Bar`s wouldn't require custom `Cell`s or a custom `Legend` `payload`. However, it doesn't stack the bars in single rows.
      - `stackId` doesn't work with range data, since it changes both the x and y positioning.

This seems to be a duplicate of https://github.com/recharts/recharts/issues/1510, which is still open from 2018.

**Comment by ckifer** [comment]
This would be a great thing to work on after we have 3.x out. Unfortunately, anything before then will fall below the line because we're just a few folks working on things as we can and we major refactoring to do so changes like this become easier for everyone. 

If someone would like to come up with an ideal for this or even some requirements (that benefit everyone), it'll make implementation down the line easier.

**Comment by faahim** [comment]
Chiming in just to up vote this.

I've been hoping to be able to do something very similar with Recharts for quite some time now. In my case, it's a ranged bar chart with varying starting/ending positions. For example:

![IMG_20240706_090238_168.jpg](https://github.com/recharts/recharts/assets/21854687/43a09c63-61ab-4a46-a639-cf99d375c5fc)


Happy to finally see it being discussed. Looking forward to v3. :)

**Comment by PavelVanecek** [comment]
We have timeline example in storybook now: https://main--63da8268a0da9970db6992aa.chromatic.com/?path=/story/examples-barchart--timeline

**Comment by Misiu** [comment]
I know that this is closed, but before I create a new issue with a similar feature request, I'd like to know if the timeline chart can be used to create this type of chart:

<img width="1645" height="209" alt="Image" src="https://github.com/user-attachments/assets/3d33780f-2dd8-4d6f-b295-1cc5b89ad499" />

This type is used in Home Assistant, and I'd like to add a similar type using recharts

**Comment by ckifer** [comment]
Hey, I would consider that a Gantt chart which isn't supported out of the box

https://github.com/recharts/recharts/issues/2847

**Comment by Misiu** [comment]
Grafana has a nice chart type for this:

<img width="778" height="298" alt="Image" src="https://github.com/user-attachments/assets/0989422f-2cbb-4dba-86c9-b3b30623d4d3" />

https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/state-timeline/#time-series-data

This is what I have for now, mostly using ComposedChart:

<img width="1628" height="282" alt="Image" src="https://github.com/user-attachments/assets/44baf877-5c98-49b5-a80a-1acc5a2aaaaf" />

**Comment by chxrise** [comment]
@Misiu The `ComposedChart` portion looks exactly like what I need, do you happen to have that exemplar code?
